src/main/java/tel/schich/javacan/BcmMessage.java

```diff
@@ -0,0 +1,308 @@
+        this.buffer = buffer;
+        this.base = buffer.position();
+        this.size = buffer.remaining();
+
+        if (size < HEADER_LENGTH) {
+            throw new IllegalArgumentException("the buffer is too small for a BCM message");
+        }
+        int expectedSize = HEADER_LENGTH + getNFrames() * frameLength(getFlags());
+        if (expectedSize > size) {
+            throw new IllegalArgumentException(String.format(
+                    "the buffer capacity cannot hold all frames of this BCM message,required %d but was %d",
+                    expectedSize, size));
+        }
+    }
+
+    @Builder
+    private BcmMessage(BcmOpcode opcode, @Singular Set<BcmFlag> flags, int count, BcmTimeval ival1, BcmTimeval ival2,
+            int can_id, @Singular List<CanFrame> frames) {
+        Objects.requireNonNull(opcode, "opcode must not be null");
+        boolean fdFrames = frames.stream().filter(CanFrame::isFDFrame).findAny().isPresent();
+        if (fdFrames && !flags.contains(BcmFlag.CAN_FD_FRAME)) {
+            flags = new HashSet<>(flags); // the set from the builder is immutable
+            flags.add(BcmFlag.CAN_FD_FRAME);
+        }
+        int frameLength = frameLength(flags);
+        base = 0;
+        size = HEADER_LENGTH + frames.size() * frameLength;
+        buffer = ByteBuffer.allocateDirect(size);
+
+        buffer.order(ByteOrder.nativeOrder())
+                .putInt(OFFSET_OPCODE, opcode.nativeOpcode)
+                .putInt(OFFSET_FLAGS, BcmFlag.toNative(flags))
+                .putInt(OFFSET_COUNT, count)
+                .putInt(OFFSET_CAN_ID, can_id)
+                .putInt(OFFSET_NFRAMES, frames.size());
+        if (ival1 != null) {
+            putPlatformLong(OFFSET_IVAL1_TV_SEC, ival1.tv_sec);
+            putPlatformLong(OFFSET_IVAL1_TV_USEC, ival1.tv_usec);
+        }
+        if (ival2 != null) {
+            putPlatformLong(OFFSET_IVAL2_TV_SEC, ival2.tv_sec);
+            putPlatformLong(OFFSET_IVAL2_TV_USEC, ival2.tv_usec);
+        }
+        for (int i = 0; i < frames.size(); i++) {
+            buffer.position(OFFSET_FRAMES + i * frameLength);
+            buffer.put(frames.get(i).getBuffer());
+        }
+        buffer.clear();
+    }
+
+    /**
+     * Returns the op-code of this message.
+     */
+    public BcmOpcode getOpcode() {
+        return BcmOpcode.fromNative(buffer.getInt(base + OFFSET_OPCODE));
+    }
+
+    /**
+     * Returns the flags of this message.
+     */
+    public Set<BcmFlag> getFlags() {
+        return BcmFlag.fromNative(buffer.getInt(base + OFFSET_FLAGS));
+    }
+
+    /**
+     * Returns the count for ival1 repetitions of this message.
+     */
+    public int getCount() {
+        return buffer.getInt(base + OFFSET_COUNT);
+    }
+
+    /**
+     * Returns ival1 from this message.
+     *
+     * @return the {@code ival1} value or {@code null} if it is not set
+     */
+    public BcmTimeval getIval1() {
```

Comment:
While this is a low level library, I'd say we can improve the naming here (the same for other methods). The library is supposed to feel like a Java library even though it is only a thin layer over SocketCAN.

Reply:
I did a rework of the naming and JavaDOC. It feels more like Java now.

For the `ival1`, `ival2` it is hard to come up with a meaningful name as these intervals have different meanings when used with different OP codes. I placed some explanations in the JavaDoc.

Reply:
if there is nothing more meaningful, at least write out ival as interval then?

Reply:
i did.